plan/new_physical_plan_builder.go

```diff
@@ -101,172 +101,186 @@ func joinKeysMatchIndex(keys []*expression.Column, index *model.IndexInfo) []int
+	lProp := &requiredProp{taskTp: rootTaskType, cols: []*expression.Column{leftKey}}
+	rProp := &requiredProp{taskTp: rootTaskType, cols: []*expression.Column{rightKey}}
+	if !prop.isEmpty() {
+		if !prop.equal(lProp) && !prop.equal(rProp) {
+			return nil
+		}
+		if prop.equal(rProp) && p.JoinType == LeftOuterJoin {
+			return nil
+		}
+		if prop.equal(lProp) && p.JoinType == RightOuterJoin {
+			return nil
+		}
+	}
+
+	return [][]*requiredProp{{lProp, rProp}}
+}
+
+func (p *LogicalJoin) tryToGetIndexJoin() ([]PhysicalPlan, bool) {
+	if len(p.EqualConditions) == 0 {
+		return nil, false
+	}
+	plans := make([]PhysicalPlan, 0, 2)
 	leftOuter := (p.preferINLJ & preferLeftAsOuter) > 0
 	if leftOuter {
 		if p.JoinType != RightOuterJoin {
-			bestTask, err = p.convertToIndexJoin(prop, 0)
-			if err != nil {
-				return nil, errors.Trace(err)
+			join := p.getIndexJoinByOuterIdx(0)
+			if join != nil {
+				plans = append(plans, join)
 			}
 		}
 	}
 	rightOuter := (p.preferINLJ & preferRightAsOuter) > 0
-	if rightOuter {
-		if p.JoinType != LeftOuterJoin {
-			task, err := p.convertToIndexJoin(prop, 1)
-			if err != nil {
-				return nil, errors.Trace(err)
-			}
-			if bestTask == nil || bestTask.cost() > task.cost() {
-				bestTask = task
-			}
+	if rightOuter && p.JoinType != LeftOuterJoin {
+		join := p.getIndexJoinByOuterIdx(1)
+		if join != nil {
+			plans = append(plans, join)
 		}
 	}
-	return
-}
-
-// convert2NewPhysicalPlan implements PhysicalPlan interface.
-// Join has three physical operators: Hash Join, Merge Join and Index Look Up Join. We implement Hash Join at first.
-func (p *LogicalJoin) convert2NewPhysicalPlan(prop *requiredProp) (task, error) {
-	task, err := p.getTask(prop)
-	if err != nil {
-		return nil, errors.Trace(err)
+	if len(plans) > 0 {
+		return plans, true
 	}
-	if task != nil {
-		return task, nil
+	if !leftOuter && p.JoinType != RightOuterJoin { // try left
+		join := p.getIndexJoinByOuterIdx(0)
+		if join != nil {
+			plans = append(plans, join)
+		}
 	}
-	if prop.taskTp != rootTaskType {
-		// Join cannot be pushed down currently, it can only return rootTask.
-		return invalidTask, p.storeTask(prop, invalidTask)
+	if !rightOuter && p.JoinType != LeftOuterJoin {
+		join := p.getIndexJoinByOuterIdx(1)
+		if join != nil {
+			plans = append(plans, join)
+		}
 	}
+	return plans, false
+}
+
+func (p *LogicalJoin) generatePhysicalPlans() []PhysicalPlan {
 	switch p.JoinType {
 	case SemiJoin, LeftOuterSemiJoin:
-		task, err = p.convert2SemiJoin(prop)
+		return []PhysicalPlan{p.getSemiJoin()}
 	default:
+		mj := p.getMergeJoin()
 		if p.preferUseMergeJoin() {
-			task, err = p.convert2MergeJoin(prop)
-		} else if task, err = p.tryToGetIndexJoin(prop); task == nil && err == nil {
-			task, err = p.convert2HashJoin(prop)
+			return []PhysicalPlan{mj}
 		}
+		joins := make([]PhysicalPlan, 0, 5)
+		if len(p.EqualConditions) == 1 {
+			joins = append(joins, mj)
+		}
+		idxJoins, forced := p.tryToGetIndexJoin()
```

Comment:
Yes. So If the `forced` is false, we will append HashJoin to possible physical plans.